packages/common/container-definitions/src/audience.ts

```diff
@@ -67,4 +73,24 @@ export interface IAudience extends EventEmitter {
 	 * does not technically have a clientId tied to an active connection to the service.
 	 */
 	getMember(clientId: string): IClient | undefined;
+
+	/**
+	 * Returns this client's clientId. undefined if this client has never connected to the ordering service.
+	 * @experimental
+	 * 
+	 * @remarks
+	 * This API is experimental.
+	 *
+	 * Whenever this property changes, the "clientIdChanged" event is fired on this object.
+	 * It's guaranteed that the "connected" event that fires at various layers, "clientIdChanged" event on this object, and the change of current clientId, all happen at the same time (synchronously, one after another).
+	 * That said, at the moment this is an experimental API. It depends on some experimental settings that might change in the future.
+	 * And application that deploy loader & container runtime bundles independently will see new (synchronized) behavior only when loader changes are deployed.
+	 * Newer runtimes will continue to observe old (non-synchronized) behavior when paired with older loader code.
+	 *
+	 * While it's marked as experimental, this promise could be broken, and consumers could experience current clientId being changed
+	 * (and "clientIdChanged" event fired) while (only applicable for "read" kind of connections)
+	 * 1. Such clientId is not present in Audience
+	 * 2. Client is not fully caught up
+	 */
+	readonly currentClientId: string | undefined;
```

Comment:
I'm not really a fan of this direction - clientId is a confusing concept for our customers so I don't like pushing it further on them.  Probably prefer to figure out what scenarios they would need the clientId for and instead provide direct solutions for those scenarios.

ServiceAudience went a direction of providing `getMyself()` which I think is better for the scenarios it covers.  In general, I'd like to see changes to Audience matched with changes to ServiceAudience to help converge, whereas this change increases divergence between the two.

Reply:
getMyself() does not help here, as there is no event to signify when it changes.

And that's what users explicitly asking for - see two Teams thread references in https://dev.azure.com/fluidframework/internal/_workitems/edit/6364

See also Tony's reply in same ADO item:
> i might narrow the scenario here, and say, a consumer of fluid should be able to access its client's details in the audience if they exist, and they should be able to determine when those details come into existence. 
i would probably explore exposing something like this directly on the audience itself, and decouple if from other system level information about connection, and client ids.

Part of the problem on why they need it - we lie to users about connections. If user has read-only permission, we never propagate connection signals to container runtime, as in reality setConnectionState() is not about connections, it's a signal to container runtime "now you can send ops". We need to start breaking this coupling of when clientId changes (and how we expose it) to other things.

I'm happy to look at ServiceAudience, but likely this needs to be a separate PR / discussion.

Reply:
My read of that thread is that they're basically trying to getMyself() - the questions about clientId are just because they need it to index into the Audience to get what they need (same scenario that getMyself() was trying to solve).

Point taken that there might be some missing eventing, and that the connection stuff is confusing, but I'm at least not seeing where the clientId itself is semantically meaningful to the customer.

Reply:
We can do it, but I'd rather do it as a layer above Audience. And as part of it - figure out higher-level API.
I'd prefer to do it as add-on layer for multiple reasons:
1. It's hard to make changes in Loader layer - they take forever to propagate, and managing different API across versions is very painful for clients (they can't use it until Loader changes saturate in production).
2. It could be expensive (if we need to rebuild state how Azure does), and it's better be pay-per-play - only used where it's needed.

We can go small step forward here and expose Self() that returns IClient & clientId, but to be honest - I'm not sure it adds that much value - anybody can fetch IClient from audience if needed.
Loop, for example, manages presence per clientId. There are other examples where people build various tracking mechanisms based on clientId (including our own DDSs). I'd rather invest into that next layer and think through what "self" should be (somewhat similar to Azure layer), i.e. it needs to be more than IClient in order to be truly useful IMHO.

Reply:
I've pushed latest commit that exposes self() as a pair of clientId & IClient.
I'd say one positive thing - it forced a bit more English to appear in comments.
Take a look.